src/runtime-rs/crates/resource/src/network/endpoint/physical_endpoint.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::path::Path;
+
+use anyhow::{Context, Result};
+use async_trait::async_trait;
+use hypervisor::{device, Hypervisor};
+
+use super::Endpoint;
+use crate::network::utils::{self, link};
+
+pub const SYS_PCI_DEVICES_PATH: &str = "/sys/bus/pci/devices";
+
+#[derive(Debug)]
+pub struct VendorDevice {
+    vendor_id: String,
+    device_id: String,
+}
+
+impl VendorDevice {
+    pub fn new(vendor_id: &str, device_id: &str) -> Self {
+        Self {
+            vendor_id: vendor_id.to_string(),
+            device_id: device_id.to_string(),
+        }
+    }
+
+    pub fn vendor_device_id(&self) -> String {
+        let vendor_device_id = format!("{} {}", &self.vendor_id, &self.device_id);
+        String::from(vendor_device_id.trim())
+    }
+}
+
+#[derive(Debug)]
+pub struct PhysicalEndpoint {
+    iface_name: String,
+    hard_addr: String,
+    bdf: String,
+    driver: String,
+    vendor_device_id: VendorDevice,
+}
+
+impl PhysicalEndpoint {
+    pub fn new(name: &str, hardware_addr: &[u8]) -> Result<Self> {
+        let driver_info = link::get_driver_info(name).context("get driver info")?;
+        let bdf = driver_info.bus_info;
+        let sys_pci_devices_path = Path::new(SYS_PCI_DEVICES_PATH);
+        // get driver by following symlink /sys/bus/pci/devices/$bdf/driver
+        let driver_path = sys_pci_devices_path.join(&bdf).join("driver");
+        let link = driver_path.read_link().context("read link")?;
+        let driver = link
+            .file_name()
+            .map_or(String::new(), |v| v.to_str().unwrap().to_owned());
+
+        // get vendor and device id from pci space (sys/bus/pci/devices/$bdf)
+        let iface_device_path = sys_pci_devices_path.join(&bdf).join("device");
+        let contents = std::fs::read(&iface_device_path)
+            .context(format!("read device path {:?}", &iface_device_path))?;
+        let device_id = String::from_utf8(contents).context("device id from utf8")?;
+
+        let iface_vendor_path = sys_pci_devices_path.join(&bdf).join("vendor");
+        let contents = std::fs::read(&iface_vendor_path)
+            .context(format!("read vendor path {:?}", &iface_vendor_path))?;
+        let vendor_id = String::from_utf8(contents).context("vendor id from utf8")?;
```

Comment:
I think it would be better if we can use fs::read_to_string() directly.

Reply:
Thanks for your review, use read_to_string directly.